### ISSUE TITLE
05 How do I use arrays in C++

### ISSUE THREAD

**Issue by zwkcoding** [zwkcoding/100Days-Of-Cplusplus-bash#5]
refer to :
https://stackoverflow.com/questions/4810664/how-do-i-use-arrays-in-c
https://stackoverflow.com/questions/24136120/what-is-wrong-with-using-arrays-dynamically-allocated-in-c?noredirect=1&lq=1

**Comment by zwkcoding** [comment]
- [ ] [Shell Style Guide](https://google.github.io/styleguide/shell.xml)

**Comment by zwkcoding** [comment]
### not to use dynamic arrays
> I heard that such use (not this code precisely, but dynamic allocation as a whole) can be unsafe in some cases, and should be used only with RAII. Why?

Take this example (similar to yours):

```
int f()
{
    char *local_buffer = new char[125];
    get_network_data(local_buffer);
    int x = make_computation(local_buffer);
    delete [] local_buffer;
    return x;
}
```

This is trivial.

Even if you write the code above correctly, somebody may come one year later, and add a conditional, or ten or twenty, in your function:

```
int f()
{
    char *local_buffer = new char[125];
    get_network_data(local_buffer);
    int x = make_computation(local_buffer);
    if(x == 25)
    {
        delete[] local_buffer;
        return 2;
    }
    if(x < 0)
    {
        delete[] local_buffer; // oops: duplicated code
        return -x;
    }
    if(x || 4)
    {
        return x/4; // oops: developer forgot to add the delete line
    }
    delete[] local_buffer; // triplicated code
    return x;
}
```
TLDR: Using raw pointers in C++ for memory management is a bad practice (altough for implementing an observer role, an implementation with raw pointers, is fine). Resource management with raw poiners violates [SRP](http://en.wikipedia.org/wiki/Single_responsibility_principle) and [DRY](http://en.wikipedia.org/wiki/Don't_repeat_yourself) principles).

**Comment by zwkcoding** [comment]
ref to : https://stackoverflow.com/questions/6500313/why-should-c-programmers-minimize-use-of-new
Reasons to use dynamic allocation:
- You don't know how much memory you need at compile time. 
- You want to allocate memory which will persist after leaving the current block.`string readfile(string path)`
Why you should use automatic storage as often as possible
### Bonus points

In the referenced question, there are additional concerns. In particular, the following class:

```
class Line {
public:
    Line();
    ~Line();
    std::string* mString;
};

Line::Line() {
    mString = new std::string("foo_bar");
}

Line::~Line() {
    delete mString;
}
```

Is actually a lot more risky to use than the following one:

```
class Line {
public:
    Line();
    std::string mString;
};

Line::Line() {
    mString = "foo_bar";
    // note: there is a cleaner way to write this.
}
```

The reason is that `std::string` properly defines a copy constructor. Consider the following program:

```
int main ()
{
    Line l1;
    Line l2 = l1;
}
```

Using the original version, this program will likely crash, as it uses `delete` on the same string twice. Using the modified version, each `Line` instance will own its own string *instance*, each with its own memory and both will be released at the end of the program.

**Comment by zwkcoding** [comment]
ref to : https://stackoverflow.com/questions/3428750/memory-leak-with-stdstring-when-using-stdliststdstring

In the first case, the `list` class has no idea you allocated the string with `new`, and cannot delete it. In particular, the list only ever contains a copy of the string that you passed in

Alternatively, just have a `std::string` member rather than a pointer and let the `string` class handle the memory (that's what it's for).

any dynamic allocation performed in a class implies you need to implement a copy-constructor and assignment operator. When you push onto a list, it pushes a copy of the object. With the code as is, there will be two instances of the object containing a pointer addressing the same location; so you could be running the risk of deleting the string twice.
